core/src/main/scala/ai/h2o/sparkling/backend/converters/DataTypeConverter.scala

```diff
@@ -0,0 +1,76 @@
+/*
+* Licensed to the Apache Software Foundation (ASF) under one or more
+* contributor license agreements.  See the NOTICE file distributed with
+* this work for additional information regarding copyright ownership.
+* The ASF licenses this file to You under the Apache License, Version 2.0
+* (the "License"); you may not use this file except in compliance with
+* the License.  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+package ai.h2o.sparkling.backend.converters
+
+import ai.h2o.sparkling.extensions.serde.ChunkSerdeConstants
+import org.apache.spark.ExposeUtils
+import org.apache.spark.h2o.utils.SupportedTypes
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+import water.fvec.Vec
+import water.parser.{BufferedString, PreviewParseWriter}
+
+object DataTypeConverter {
+  private def stringTypesToExpectedTypes(rdd: RDD[Row], schema: StructType): Map[Int, Byte] = {
+    val stringTypeIndices = for {
+      (field, index) <- schema.fields.zipWithIndex
+      if field == StringType
+    } yield index
+
+    val preview = rdd
+      .mapPartitions[PreviewParseWriter](createPartitionPreview(_, stringTypeIndices))
+      .reduce(PreviewParseWriter.unifyColumnPreviews)
+
+    val types = preview.guessTypes().map {
+      case Vec.T_STR => ChunkSerdeConstants.EXPECTED_STRING
+      case Vec.T_CAT => ChunkSerdeConstants.EXPECTED_CATEGORICAL
+    }
+
+    stringTypeIndices.zip(types).toMap
+  }
+
+  private def createPartitionPreview(rows: Iterator[Row], stringTypeIndices: Array[Int]): Iterator[PreviewParseWriter] = {
+    val previewParseWriter = new PreviewParseWriter()
```

Comment:
Preview parse writer has package protected constructor. We can either change on H2O side or create subclass from this class on SW side with public constructor

Reply:
Good catch, thanks! I prefer to create a subclass here.